File: leetcode/queue_circular.py
#!/usr/bin/env python3

import sys
import unittest

# number: 622
# section: citadel
# difficulty: medium
# tags: array, linked list, design, queue, citadel

# constraints
# 1 <= k <= 1000
# 0 <= value <= 1000
# At most 3000 calls will be made to enqueue, dequeue, front, rear, isempty, and
# isfull.

# solution: 2 Python lists/arrays using append, front & rear
# complexity
# run-time: O(1) per op
# space: O(k)
class MyCircularQueue:

    def __init__(self, k: int):
        self.__front = []
        self.__rear = []
        self.__capacity = k

    def enqueue(self, value: int) -> bool:
        if self.isfull():
            return False

        self.__rear.append(value)

        return True

    def dequeue(self) -> bool:
        if self.__front:
            del self.__front[0]
            return True
        elif self.__rear:
            del self.__rear[0]
            return True
        else:
            return False

    def front(self) -> int:
        # first check front!
        if self.__front:
            return self.__front[0]
        elif self.__rear:
            return self.__rear[0]
        else:
            return -1

    def rear(self) -> int:
        # first check rear!
        if self.__rear:
            return self.__rear[-1]
        elif self.__front:
            return self.__front[-1]
        else:
            return -1

    def isempty(self) -> bool:
        if self.__front or self.__rear:
            return False

        return True

    def isfull(self) -> bool:
        if len(self.__front) + len(self.__rear) == self.__capacity:
            return True

        return False

# solution: 1 Python list w/o append
# + calc tail: (head + size-1) % capacity
# complexity
# run-time: O(1) per op
# space: O(k)
class MyCircularQueue2:

    def __init__(self, k: int):
        # allocate memory: no appends!
        self.__queue = [None]*k
        self.__capacity = k
        self.__size = 0
        self.__head = 0

    def enqueue(self, value: int) -> bool:
        # TODO: add increase capacity by doubling?
        if self.isfull():
            return False

        # don't subtract 1: adding a new item!
        self.__queue[(self.__head + self.__size) % self.__capacity] = value
        self.__size += 1

        return True

    def dequeue(self) -> bool:
        if self.isempty():
            return False

        # make circular: if past end!
        self.__head = (self.__head + 1) % self.__capacity
        self.__size -= 1

        return True

    def front(self) -> int:
        if self.isempty():
            return -1

        return self.__queue[self.__head]

    def rear(self) -> int:
        if self.isempty():
            return -1

        # calc tail
        return self.__queue[(self.__head + self.__size-1) % self.__capacity]

    def isempty(self) -> bool:
        return not self.__size

    def isfull(self) -> bool:
        return self.__size == self.__capacity

# TODO: solve using linked list

class TestMyCirciularQueue(unittest.TestCase):
    def test1(self):
        k = 3
        obj = MyCircularQueue(k)
        self.assertTrue(obj.enqueue(1))
        self.assertTrue(obj.enqueue(2))
        self.assertTrue(obj.enqueue(3))
        self.assertFalse(obj.enqueue(4))
        self.assertEqual(obj.rear(), 3)
        self.assertTrue(obj.isfull())
        self.assertTrue(obj.dequeue())
        self.assertTrue(obj.enqueue(4))
        self.assertEqual(obj.rear(), 4)
        self.assertFalse(obj.isempty())
        self.assertEqual(obj.front(), 2)

    def test2(self):
        k = 3
        obj = MyCircularQueue2(k)
        self.assertTrue(obj.enqueue(1))
        self.assertTrue(obj.enqueue(2))
        self.assertTrue(obj.enqueue(3))
        self.assertFalse(obj.enqueue(4))
        self.assertEqual(obj.rear(), 3)
        self.assertTrue(obj.isfull())
        self.assertTrue(obj.dequeue())
        self.assertTrue(obj.enqueue(4))
        self.assertEqual(obj.rear(), 4)
        self.assertFalse(obj.isempty())
        self.assertEqual(obj.front(), 2)

if __name__ == '__main__':
    sys.exit(unittest.main())

File: leetcode/queue_circular.cpp
#include <vector>
#include <stdexcept>

// number: 622
// title: Design Circular Queue
// url: https://leetcode.com/problems/design-circular-queue/
// section: citadel
// difficulty: medium
// tags: array, linked list, design, queue, citadel
//
// constraints
// 1 <= k <= 1000
// 0 <= value <= 1000
// At most 3000 calls will be made to enqueue, dequeue, front, rear, isempty, and
// isfull.

// solution: 2 vectors, using push_back
// complexity
// run-time: O(1) per op
// space: O(k)
template <typename T>
class MyCircularQueue {
public:
    MyCircularQueue(int k)
    : d_capacity(k)
    {
        if (k <= 0) {
            throw std::invalid_argument("invalid capacity");
        }

        // don't use reserve(), don't know how to split it b/w front & rear
    }

    // don't allow creation of queue with 0 capacity
    MyCircularQueue() = delete;

    // don't support copying
    MyCircularQueue(const MyCircularQueue&) = delete;
    MyCircularQueue& operator=(const MyCircularQueue&) = delete;

    // don't support moving
    MyCircularQueue(MyCircularQueue&&) = delete;
    MyCircularQueue& operator=(MyCircularQueue&&) = delete;

    bool enQueue(const T& value) {
        if (isFull()) {
            return false;
        }

        d_rear.push_back(value);
        return true;
    }

    bool deQueue() {
        if (isEmpty()) {
            return false;
        }

        if (!d_front.empty()) {
            d_front.erase(d_front.begin());
            return true;
        } else if (!d_rear.empty()) {
            d_rear.erase(d_rear.begin());
            return true;
        }

        return false;
    }

    T Front() {
        // first check front!
        if (!d_front.empty()) {
            return d_front.front();
        } else if (!d_rear.empty()) {
            return d_rear.front();
        }

        throw std::length_error("empty queue");
    }

    T Rear() {
        // first check rear!
        if (!d_rear.empty()) {
            return d_rear.back();
        } else if(!d_front.empty()) {
            return d_front.back();
        }

        throw std::length_error("empty queue");
    }

    bool isEmpty() {
        if (!d_front.empty() or !d_rear.empty()) {
            return false;
        }

        return true;
    }

    bool isFull() {
        if (d_front.size() + d_rear.size() >= d_capacity) {
            return true;
        }

        return false;
    }

private:
    std::vector<T> d_front;
    std::vector<T> d_rear;
    int d_capacity;
};

// solution: 1 vector w/o push_back
// + calc tail: (head + size-1) % capacity
// complexity
// run-time: O(1) per op
// space: O(k)
const size_t MAX_SIZE = 100;

template <typename T>
class MyCircularQueue2 {
public:
    MyCircularQueue2()
    : d_capacity(MAX_SIZE)
    , d_size(0)
    , d_head(0)
    {
        d_queue.reserve(d_capacity);
    }

    MyCircularQueue2(int k)
    : d_capacity(k)
    , d_size(0)
    , d_head(0)
    {
        if (k <= 0) {
            throw std::invalid_argument("invalid capacity");
        }

        // allocate memory
        d_queue.reserve(d_capacity);
    }

    // don't support copying
    MyCircularQueue2(const MyCircularQueue2&) = delete;
    MyCircularQueue2& operator=(const MyCircularQueue2&) = delete;

    // don't support moving
    MyCircularQueue2(MyCircularQueue2&&) = delete;
    MyCircularQueue2& operator=(MyCircularQueue2&&) = delete;

    bool enQueue(const T& value) {
        // TODO: increase capacity by doubling?
        if (isFull()) {
            return false;
        }

        // don't subtract 1: adding a new item!
        d_queue[(d_head + d_size) % d_capacity] = value;
        ++d_size;

        return true;
    }

    bool deQueue() {
        if (isEmpty()) {
            return false;
        }

        // make circular: if past end!
        d_head = (d_head + 1) % d_capacity;
        --d_size;

        return true;
    }

    T Front() {
        if (isEmpty()) {
            throw std::length_error("empty queue");
        }

        return d_queue[d_head];
    }

    T Rear() {
        if (isEmpty()) {
            throw std::length_error("empty queue");
        }

        // calc tail
        return d_queue[(d_head + d_size-1) % d_capacity];
    }

    bool isEmpty() { return d_size == 0; }

    bool isFull() { return d_size == d_capacity; }

private:
    std::vector<T> d_queue;
    size_t         d_capacity;
    size_t         d_size;
    size_t         d_head;
};

// TODO: solve using linked list & unit test

// Your MyCircularQueue object will be instantiated and called as such:
MyCircularQueue* obj = new MyCircularQueue(k);
bool param_1 = obj->enQueue(value);
bool param_2 = obj->deQueue();
int param_3 = obj->Front();
int param_4 = obj->Rear();
bool param_5 = obj->isEmpty();
bool param_6 = obj->isFull();
